Replacing login with request to Cognito

HeyZoos opened the pull request:
Updates the login function to authenticate via Cognito, and provide the access token returned from Cognito to the Pennsieve backend.

The Rust client is using a very out-of-date version of the futures library, which was before async programming stabilized in the language. I tried to bring it up to date but it proved beyond me. I still think it's important to do however since I think we'll keep running into compatibility errors with packages that use the more up to date futures. For example, I had to revert the Cognito client to an older version to get around the future type incompatibility.

HeyZoos committed:
[skip-ci] Start replacing login with request to Cognito

HeyZoos committed:
Check in, incompatability with future types

HeyZoos committed:
Use a rusoto client compatible with futures v1

HeyZoos committed:
Revert `get_pennsieve_url` to `get_url`

HeyZoos committed:
Get most of the login tests passing

rlmv commented:
@mikeswoods Would you mind reviewing this when you have some time?

rlmv commented on src/ps/api/client/mod.rs:
```diff
@@ -1217,6 +1243,7 @@ pub mod tests {
     const TEST_ENVIRONMENT: Environment = Environment::NonProduction;
     const TEST_API_KEY: &str = env!("PENNSIEVE_API_KEY");
     const TEST_SECRET_KEY: &str = env!("PENNSIEVE_SECRET_KEY");
+    const TEST_API_ID: &str = env!("PENNSIEVE_API_ID");
```
I don't love that we have this new `PENNSIEVE_API_ID` that the clients need to know about in order to authenticate. What would you think about exposing a new un-authenticated endpoint that serves up the Cognito client and pool IDs? Then the clients could hit that endpoint when they first start up and grab to the token pool client ID.

Otherwise we are going to have to configure these client IDs all over the place for the different environments. Would be much cleaner if the clients could figure this out themselves.

HeyZoos replied:
Yeah I think could work 👍

rlmv replied:
That shouldn't block merging this, but let's try and clean that up soon.

rlmv replied:
Ticket here to track: https://app.clickup.com/t/rd87ae

mikeswoods commented on src/ps/api/client/mod.rs:
```diff
@@ -547,18 +554,37 @@ impl Pennsieve {
+            user_context_data: None,
+        };
+
+        let mut runtime = tokio::runtime::current_thread::Runtime::new().unwrap();
+        let response = runtime.block_on(cognito.initiate_auth(request)).unwrap();
```
If this method is returning a `Future`, why is it necessary to block in the `Future` body? Why not await the result of the `initiate_auth`(same as `and_then`), and then set the `session_token` to that of the access_token value? I very well could be mistaken since I haven't touched Rust in a while!

I get it if it's needed to get things to work (a common Rust problem!) If so, emitting a compiler warning and including a comment would suffice to warn future users that the `Future` returned by `login` will block the current thread, as it is unexpected behavior.

Also, I know the version of `tokio` that's being used is ancient at this point, but it's not necessarily safe to assume that the executor driving the chain of futures is the same implementation of `current_thread`. You can encounter race conditions when mixing executors.

HeyZoos replied:
Yeah it was definitely a "get it working" hack. The `0.43` version of `initiate_auth` function returns this `Pin<Box<dyn Future<Output = Result<InitiateAuthResponse, RusotoError<InitiateAuthError>>> + Send + 'async_trait>>`. I'm not comfortable with working with Futures and was having trouble unwrapping the types.

I'll definitely add a warning and I'll look into the point your raise about the executor. Thanks for the comprehensive review. 🙏

mikeswoods commented:
@HeyZoos Thanks for tackling this. The version of `future` used for this is pretty old as you pointed out. We'd be better of rewriting the library with async+await support plus the async version of `reqwest` (https://github.com/seanmonstar/reqwest). I never liked using `hyper` as the basis for the `blackfynn-rust` library (it felt too low-level), but `reqwest` was still in its infancy and did not support async execution. 

As a stop gap, I guess do whatever you need to do to get it working with Cognito. At some point in the future someone will have to rewrite this.

HeyZoos committed:
Retrive cognito config on the fly

HeyZoos committed:
Extract and use organization id

HeyZoos committed:
Get rid of `jsonwebtoken` dep

HeyZoos committed:
Revert `pennsieve_url` to `url`

HeyZoos committed:
Remove `dbg!`

HeyZoos committed:
Add expiration info

rlmv committed:
Use default static credentials provider

This credentials provider does not look for AWS keys in the
environment. Should fix test failures on Github.

rlmv commented:
@HeyZoos Tests are now working on CI with a static / default credential provider. I am going to go ahead and merge this to try and push out the Agent.
